core/src/subgraph/registrar.rs

```diff
@@ -0,0 +1,776 @@
+                                    "subgraph_id" => subgraph_id.to_string()
+                                );
+                                Ok(())
+                            }
+                        }
+                    }),
+            )
+        }
+        DeploymentEvent::Remove {
+            subgraph_id,
+            node_id: _,
+        } => Box::new(
+            provider
+                .stop(subgraph_id)
+                .then(|result| match result {
+                    Ok(()) => Ok(()),
+                    Err(SubgraphDeploymentProviderError::NotRunning(_)) => Ok(()),
+                    Err(e) => Err(e),
+                })
+                .map_err(CancelableError::Error),
+        ),
+    }
+}
+
+fn create_subgraph(
+    logger: &Logger,
+    store: Arc<impl Store>,
+    name: SubgraphName,
+) -> Result<(), SubgraphRegistrarError> {
+    debug!(logger, "Creating subgraph: {:?}", name.to_string());
+
+    let mut ops = vec![];
+
+    // Check if this subgraph already exists
+    let subgraph_entities = store.find(SubgraphEntity::query().filter(EntityFilter::Equal(
+        "name".to_owned(),
+        name.to_string().into(),
+    )))?;
+    if !subgraph_entities.is_empty() {
+        return Err(SubgraphRegistrarError::NameExists(name.to_string()));
+    }
+
+    ops.push(EntityOperation::AbortUnless {
+        description: "Subgraph entity should not exist".to_owned(),
+        query: SubgraphEntity::query().filter(EntityFilter::Equal(
+            "name".to_owned(),
+            name.to_string().into(),
+        )),
+        entity_ids: vec![],
+    });
+
+    let created_at = SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .unwrap()
+        .as_secs();
+    let entity = SubgraphEntity::new(name, None, created_at);
+    let entity_id = generate_entity_id();
+    ops.extend(entity.write_operations(&entity_id));
+
+    store.apply_entity_operations(ops, EventSource::None)?;
+
+    Ok(())
+}
+
+fn create_subgraph_version(
+    logger: &Logger,
+    store: Arc<impl Store>,
+    chain_store: Arc<impl ChainStore>,
+    name: SubgraphName,
+    manifest: SubgraphManifest,
+    node_id: NodeId,
+) -> Result<(), SubgraphRegistrarError> {
+    debug!(
+        logger,
+        "Writing new subgraph version to store: subgraph name = {:?}, subgraph hash = {:?}",
+        name.to_string(),
+        manifest.id.to_string()
+    );
+
+    let mut ops = vec![];
+
+    // Look up subgraph entity by name
+    let subgraph_entities = store.find(SubgraphEntity::query().filter(EntityFilter::Equal(
+        "name".to_owned(),
+        name.to_string().into(),
+    )))?;
+    let subgraph_entity = match subgraph_entities.len() {
+        0 => Err(SubgraphRegistrarError::NameNotFound(name.to_string())),
+        1 => {
+            let mut subgraph_entities = subgraph_entities;
+            Ok(subgraph_entities.pop().unwrap())
+        }
+        _ => panic!(
+            "multiple subgraph entities with name {:?}",
```

Comment:
Ah, this is just a trick to get the quotes but also to get the escaping of special characters. I've found it helpful for debugging to have special characters escaped in a predictable way. Although, I guess that doesn't matter here because of subgraph name validation...